File: sample/go/07_game_of_life_loop.go
package main


// 07: Sample that outputs Game of Life evolution as a GIF.

func next_state(grid []any, w int64, h int64) []any {
    var nxt []any = __pytra_as_list([]any{})
    for y := int64(0); y < h; y += 1 {
        var row []any = __pytra_as_list([]any{})
        for x := int64(0); x < w; x += 1 {
            var cnt int64 = int64(0)
            for dy := (-int64(1)); dy < int64(2); dy += 1 {
                for dx := (-int64(1)); dx < int64(2); dx += 1 {
                    if ((dx != int64(0)) || (dy != int64(0))) {
                        var nx int64 = (((x + dx) + w) % w)
                        var ny int64 = (((y + dy) + h) % h)
                        cnt += __pytra_int(__pytra_get_index(__pytra_as_list(__pytra_get_index(grid, ny)), nx))
                    }
                }
            }
            var alive int64 = __pytra_int(__pytra_get_index(__pytra_as_list(__pytra_get_index(grid, y)), x))
            if ((alive == int64(1)) && ((cnt == int64(2)) || (cnt == int64(3)))) {
                row = append(row, int64(1))
            } else {
                if ((alive == int64(0)) && (cnt == int64(3))) {
                    row = append(row, int64(1))
                } else {
                    row = append(row, int64(0))
                }
            }
        }
        nxt = append(nxt, row)
    }
    return __pytra_as_list(nxt)
}

func render(grid []any, w int64, h int64, cell int64) []any {
    var width int64 = (w * cell)
    var height int64 = (h * cell)
    var frame []any = __pytra_as_list(__pytra_bytearray((width * height)))
    for y := int64(0); y < h; y += 1 {
        for x := int64(0); x < w; x += 1 {
            var v int64 = __pytra_ifexp((__pytra_int(__pytra_get_index(__pytra_as_list(__pytra_get_index(grid, y)), x)) != 0), int64(255), int64(0))
            for yy := int64(0); yy < cell; yy += 1 {
                var base int64 = ((((y * cell) + yy) * width) + (x * cell))
                for xx := int64(0); xx < cell; xx += 1 {
                    __pytra_set_index(frame, (base + xx), v)
                }
            }
        }
    }
    return __pytra_as_list(__pytra_bytes(frame))
}

func run_07_game_of_life_loop() {
    var w int64 = int64(144)
    var h int64 = int64(108)
    var cell int64 = int64(4)
    var steps int64 = int64(105)
    var out_path string = __pytra_str("sample/out/07_game_of_life_loop.gif")
    var start float64 = __pytra_perf_counter()
    var grid []any = __pytra_as_list(func() []any { __out := []any{}; __step := __pytra_int(int64(1)); for __lc_i := __pytra_int(int64(0)); (__step >= 0 && __lc_i < __pytra_int(h)) || (__step < 0 && __lc_i > __pytra_int(h)); __lc_i += __step { __out = append(__out, __pytra_list_repeat(int64(0), w)) }; return __out }())
    for y := int64(0); y < h; y += 1 {
        for x := int64(0); x < w; x += 1 {
            var noise int64 = (((((x * int64(37)) + (y * int64(73))) + ((x * y) % int64(19))) + ((x + y) % int64(11))) % int64(97))
            if (noise < int64(3)) {
                __pytra_set_index(__pytra_as_list(__pytra_get_index(grid, y)), x, int64(1))
            }
        }
    }
    var glider []any = __pytra_as_list([]any{[]any{int64(0), int64(1), int64(0)}, []any{int64(0), int64(0), int64(1)}, []any{int64(1), int64(1), int64(1)}})
    var r_pentomino []any = __pytra_as_list([]any{[]any{int64(0), int64(1), int64(1)}, []any{int64(1), int64(1), int64(0)}, []any{int64(0), int64(1), int64(0)}})
    var lwss []any = __pytra_as_list([]any{[]any{int64(0), int64(1), int64(1), int64(1), int64(1)}, []any{int64(1), int64(0), int64(0), int64(0), int64(1)}, []any{int64(0), int64(0), int64(0), int64(0), int64(1)}, []any{int64(1), int64(0), int64(0), int64(1), int64(0)}})
    __step_0 := int64(18)
    for gy := int64(8); (__step_0 >= 0 && gy < (h - int64(8))) || (__step_0 < 0 && gy > (h - int64(8))); gy += __step_0 {
        __step_1 := int64(22)
        for gx := int64(8); (__step_1 >= 0 && gx < (w - int64(8))) || (__step_1 < 0 && gx > (w - int64(8))); gx += __step_1 {
            var kind int64 = (((gx * int64(7)) + (gy * int64(11))) % int64(3))
            if (kind == int64(0)) {
                var ph int64 = __pytra_len(glider)
                for py := int64(0); py < ph; py += 1 {
                    var pw int64 = __pytra_len(__pytra_as_list(__pytra_get_index(glider, py)))
                    for px := int64(0); px < pw; px += 1 {
                        if (__pytra_int(__pytra_get_index(__pytra_as_list(__pytra_get_index(glider, py)), px)) == int64(1)) {
                            __pytra_set_index(__pytra_as_list(__pytra_get_index(grid, ((gy + py) % h))), ((gx + px) % w), int64(1))
                        }
                    }
                }
            } else {
                if (kind == int64(1)) {
                    var ph int64 = __pytra_len(r_pentomino)
                    for py := int64(0); py < ph; py += 1 {
                        var pw int64 = __pytra_len(__pytra_as_list(__pytra_get_index(r_pentomino, py)))
                        for px := int64(0); px < pw; px += 1 {
                            if (__pytra_int(__pytra_get_index(__pytra_as_list(__pytra_get_index(r_pentomino, py)), px)) == int64(1)) {
                                __pytra_set_index(__pytra_as_list(__pytra_get_index(grid, ((gy + py) % h))), ((gx + px) % w), int64(1))
                            }
                        }
                    }
                } else {
                    var ph int64 = __pytra_len(lwss)
                    for py := int64(0); py < ph; py += 1 {
                        var pw int64 = __pytra_len(__pytra_as_list(__pytra_get_index(lwss, py)))
                        for px := int64(0); px < pw; px += 1 {
                            if (__pytra_int(__pytra_get_index(__pytra_as_list(__pytra_get_index(lwss, py)), px)) == int64(1)) {
                                __pytra_set_index(__pytra_as_list(__pytra_get_index(grid, ((gy + py) % h))), ((gx + px) % w), int64(1))
                            }
                        }
                    }
                }
            }
        }
    }
    var frames []any = __pytra_as_list([]any{})
    for __loop_2 := int64(0); __loop_2 < steps; __loop_2 += 1 {
        frames = append(frames, render(grid, w, h, cell))
        grid = __pytra_as_list(next_state(grid, w, h))
    }
    __pytra_save_gif(out_path, (w * cell), (h * cell), frames, __pytra_grayscale_palette(), int64(4), int64(0))
    var elapsed float64 = (__pytra_perf_counter() - start)
    __pytra_print("output:", out_path)
    __pytra_print("frames:", steps)
    __pytra_print("elapsed_sec:", elapsed)
}

func main() {
    run_07_game_of_life_loop()
}

File: sample/swift/07_game_of_life_loop.swift
import Foundation


// 07: Sample that outputs Game of Life evolution as a GIF.

func next_state(grid: [Any], w: Int64, h: Int64) -> [Any] {
    var nxt: [Any] = __pytra_as_list([])
    let __step_0 = __pytra_int(Int64(1))
    var y = __pytra_int(Int64(0))
    while ((__step_0 >= 0 && y < __pytra_int(h)) || (__step_0 < 0 && y > __pytra_int(h))) {
        var row: [Any] = __pytra_as_list([])
        let __step_1 = __pytra_int(Int64(1))
        var x = __pytra_int(Int64(0))
        while ((__step_1 >= 0 && x < __pytra_int(w)) || (__step_1 < 0 && x > __pytra_int(w))) {
            var cnt: Int64 = __pytra_int(Int64(0))
            let __step_2 = __pytra_int(Int64(1))
            var dy = __pytra_int((-Int64(1)))
            while ((__step_2 >= 0 && dy < __pytra_int(Int64(2))) || (__step_2 < 0 && dy > __pytra_int(Int64(2)))) {
                let __step_3 = __pytra_int(Int64(1))
                var dx = __pytra_int((-Int64(1)))
                while ((__step_3 >= 0 && dx < __pytra_int(Int64(2))) || (__step_3 < 0 && dx > __pytra_int(Int64(2)))) {
                    if ((__pytra_int(dx) != __pytra_int(Int64(0))) || (__pytra_int(dy) != __pytra_int(Int64(0)))) {
                        var nx: Int64 = __pytra_int((__pytra_int((__pytra_int((__pytra_int(x) + __pytra_int(dx))) + __pytra_int(w))) % __pytra_int(w)))
                        var ny: Int64 = __pytra_int((__pytra_int((__pytra_int((__pytra_int(y) + __pytra_int(dy))) + __pytra_int(h))) % __pytra_int(h)))
                        cnt += __pytra_int(__pytra_getIndex(__pytra_as_list(__pytra_getIndex(grid, ny)), nx))
                    }
                    dx += __step_3
                }
                dy += __step_2
            }
            var alive: Int64 = __pytra_int(__pytra_int(__pytra_getIndex(__pytra_as_list(__pytra_getIndex(grid, y)), x)))
            if ((__pytra_int(alive) == __pytra_int(Int64(1))) && ((__pytra_int(cnt) == __pytra_int(Int64(2))) || (__pytra_int(cnt) == __pytra_int(Int64(3))))) {
                row = __pytra_as_list(row); row.append(Int64(1))
            } else {
                if ((__pytra_int(alive) == __pytra_int(Int64(0))) && (__pytra_int(cnt) == __pytra_int(Int64(3)))) {
                    row = __pytra_as_list(row); row.append(Int64(1))
                } else {
                    row = __pytra_as_list(row); row.append(Int64(0))
                }
            }
            x += __step_1
        }
        nxt = __pytra_as_list(nxt); nxt.append(row)
        y += __step_0
    }
    return nxt
}

func render(grid: [Any], w: Int64, h: Int64, cell: Int64) -> [Any] {
    var width: Int64 = __pytra_int((__pytra_int(w) * __pytra_int(cell)))
    var height: Int64 = __pytra_int((__pytra_int(h) * __pytra_int(cell)))
    var frame: [Any] = __pytra_as_list(__pytra_bytearray((__pytra_int(width) * __pytra_int(height))))
    let __step_0 = __pytra_int(Int64(1))
    var y = __pytra_int(Int64(0))
    while ((__step_0 >= 0 && y < __pytra_int(h)) || (__step_0 < 0 && y > __pytra_int(h))) {
        let __step_1 = __pytra_int(Int64(1))
        var x = __pytra_int(Int64(0))
        while ((__step_1 >= 0 && x < __pytra_int(w)) || (__step_1 < 0 && x > __pytra_int(w))) {
            var v: Int64 = __pytra_int(__pytra_ifexp((__pytra_int(__pytra_getIndex(__pytra_as_list(__pytra_getIndex(grid, y)), x)) != 0), Int64(255), Int64(0)))
            let __step_2 = __pytra_int(Int64(1))
            var yy = __pytra_int(Int64(0))
            while ((__step_2 >= 0 && yy < __pytra_int(cell)) || (__step_2 < 0 && yy > __pytra_int(cell))) {
                var base: Int64 = __pytra_int((__pytra_int((__pytra_int((__pytra_int((__pytra_int(y) * __pytra_int(cell))) + __pytra_int(yy))) * __pytra_int(width))) + __pytra_int((__pytra_int(x) * __pytra_int(cell)))))
                let __step_3 = __pytra_int(Int64(1))
                var xx = __pytra_int(Int64(0))
                while ((__step_3 >= 0 && xx < __pytra_int(cell)) || (__step_3 < 0 && xx > __pytra_int(cell))) {
                    __pytra_setIndex(frame, (__pytra_int(base) + __pytra_int(xx)), v)
                    xx += __step_3
                }
                yy += __step_2
            }
            x += __step_1
        }
        y += __step_0
    }
    return __pytra_bytes(frame)
}

func run_07_game_of_life_loop() {
    var w: Int64 = __pytra_int(Int64(144))
    var h: Int64 = __pytra_int(Int64(108))
    var cell: Int64 = __pytra_int(Int64(4))
    var steps: Int64 = __pytra_int(Int64(105))
    var out_path: String = __pytra_str("sample/out/07_game_of_life_loop.gif")
    var start: Double = __pytra_float(__pytra_perf_counter())
    var grid: [Any] = __pytra_as_list(({ () -> [Any] in var __out: [Any] = []; let __step = __pytra_int(Int64(1)); var __lc_i = __pytra_int(Int64(0)); while ((__step >= 0 && __lc_i < __pytra_int(h)) || (__step < 0 && __lc_i > __pytra_int(h))) { __out.append(__pytra_list_repeat(Int64(0), w)); __lc_i += __step }; return __out })())
    let __step_0 = __pytra_int(Int64(1))
    var y = __pytra_int(Int64(0))
    while ((__step_0 >= 0 && y < __pytra_int(h)) || (__step_0 < 0 && y > __pytra_int(h))) {
        let __step_1 = __pytra_int(Int64(1))
        var x = __pytra_int(Int64(0))
        while ((__step_1 >= 0 && x < __pytra_int(w)) || (__step_1 < 0 && x > __pytra_int(w))) {
            var noise: Int64 = __pytra_int((__pytra_int((__pytra_int((__pytra_int((__pytra_int((__pytra_int(x) * __pytra_int(Int64(37)))) + __pytra_int((__pytra_int(y) * __pytra_int(Int64(73)))))) + __pytra_int((__pytra_int((__pytra_int(x) * __pytra_int(y))) % __pytra_int(Int64(19)))))) + __pytra_int((__pytra_int((__pytra_int(x) + __pytra_int(y))) % __pytra_int(Int64(11)))))) % __pytra_int(Int64(97))))
            if (__pytra_int(noise) < __pytra_int(Int64(3))) {
                __pytra_setIndex(__pytra_as_list(__pytra_getIndex(grid, y)), x, Int64(1))
            }
            x += __step_1
        }
        y += __step_0
    }
    var glider: [Any] = __pytra_as_list([[Int64(0), Int64(1), Int64(0)], [Int64(0), Int64(0), Int64(1)], [Int64(1), Int64(1), Int64(1)]])
    var r_pentomino: [Any] = __pytra_as_list([[Int64(0), Int64(1), Int64(1)], [Int64(1), Int64(1), Int64(0)], [Int64(0), Int64(1), Int64(0)]])
    var lwss: [Any] = __pytra_as_list([[Int64(0), Int64(1), Int64(1), Int64(1), Int64(1)], [Int64(1), Int64(0), Int64(0), Int64(0), Int64(1)], [Int64(0), Int64(0), Int64(0), Int64(0), Int64(1)], [Int64(1), Int64(0), Int64(0), Int64(1), Int64(0)]])
    let __step_2 = __pytra_int(Int64(18))
    var gy = __pytra_int(Int64(8))
    while ((__step_2 >= 0 && gy < __pytra_int((__pytra_int(h) - __pytra_int(Int64(8))))) || (__step_2 < 0 && gy > __pytra_int((__pytra_int(h) - __pytra_int(Int64(8)))))) {
        let __step_3 = __pytra_int(Int64(22))
        var gx = __pytra_int(Int64(8))
        while ((__step_3 >= 0 && gx < __pytra_int((__pytra_int(w) - __pytra_int(Int64(8))))) || (__step_3 < 0 && gx > __pytra_int((__pytra_int(w) - __pytra_int(Int64(8)))))) {
            var kind: Int64 = __pytra_int((__pytra_int((__pytra_int((__pytra_int(gx) * __pytra_int(Int64(7)))) + __pytra_int((__pytra_int(gy) * __pytra_int(Int64(11)))))) % __pytra_int(Int64(3))))
            if (__pytra_int(kind) == __pytra_int(Int64(0))) {
                var ph: Int64 = __pytra_int(__pytra_len(glider))
                let __step_4 = __pytra_int(Int64(1))
                var py = __pytra_int(Int64(0))
                while ((__step_4 >= 0 && py < __pytra_int(ph)) || (__step_4 < 0 && py > __pytra_int(ph))) {
                    var pw: Int64 = __pytra_int(__pytra_len(__pytra_as_list(__pytra_getIndex(glider, py))))
                    let __step_5 = __pytra_int(Int64(1))
                    var px = __pytra_int(Int64(0))
                    while ((__step_5 >= 0 && px < __pytra_int(pw)) || (__step_5 < 0 && px > __pytra_int(pw))) {
                        if (__pytra_int(__pytra_int(__pytra_getIndex(__pytra_as_list(__pytra_getIndex(glider, py)), px))) == __pytra_int(Int64(1))) {
                            __pytra_setIndex(__pytra_as_list(__pytra_getIndex(grid, (__pytra_int((__pytra_int(gy) + __pytra_int(py))) % __pytra_int(h)))), (__pytra_int((__pytra_int(gx) + __pytra_int(px))) % __pytra_int(w)), Int64(1))
                        }
                        px += __step_5
                    }
                    py += __step_4
                }
            } else {
                if (__pytra_int(kind) == __pytra_int(Int64(1))) {
                    var ph: Int64 = __pytra_int(__pytra_len(r_pentomino))
                    let __step_6 = __pytra_int(Int64(1))
                    var py = __pytra_int(Int64(0))
                    while ((__step_6 >= 0 && py < __pytra_int(ph)) || (__step_6 < 0 && py > __pytra_int(ph))) {
                        var pw: Int64 = __pytra_int(__pytra_len(__pytra_as_list(__pytra_getIndex(r_pentomino, py))))
                        let __step_7 = __pytra_int(Int64(1))
                        var px = __pytra_int(Int64(0))
                        while ((__step_7 >= 0 && px < __pytra_int(pw)) || (__step_7 < 0 && px > __pytra_int(pw))) {
                            if (__pytra_int(__pytra_int(__pytra_getIndex(__pytra_as_list(__pytra_getIndex(r_pentomino, py)), px))) == __pytra_int(Int64(1))) {
                                __pytra_setIndex(__pytra_as_list(__pytra_getIndex(grid, (__pytra_int((__pytra_int(gy) + __pytra_int(py))) % __pytra_int(h)))), (__pytra_int((__pytra_int(gx) + __pytra_int(px))) % __pytra_int(w)), Int64(1))
                            }
                            px += __step_7
                        }
                        py += __step_6
                    }
                } else {
                    var ph: Int64 = __pytra_int(__pytra_len(lwss))
                    let __step_8 = __pytra_int(Int64(1))
                    var py = __pytra_int(Int64(0))
                    while ((__step_8 >= 0 && py < __pytra_int(ph)) || (__step_8 < 0 && py > __pytra_int(ph))) {
                        var pw: Int64 = __pytra_int(__pytra_len(__pytra_as_list(__pytra_getIndex(lwss, py))))
                        let __step_9 = __pytra_int(Int64(1))
                        var px = __pytra_int(Int64(0))
                        while ((__step_9 >= 0 && px < __pytra_int(pw)) || (__step_9 < 0 && px > __pytra_int(pw))) {
                            if (__pytra_int(__pytra_int(__pytra_getIndex(__pytra_as_list(__pytra_getIndex(lwss, py)), px))) == __pytra_int(Int64(1))) {
                                __pytra_setIndex(__pytra_as_list(__pytra_getIndex(grid, (__pytra_int((__pytra_int(gy) + __pytra_int(py))) % __pytra_int(h)))), (__pytra_int((__pytra_int(gx) + __pytra_int(px))) % __pytra_int(w)), Int64(1))
                            }
                            px += __step_9
                        }
                        py += __step_8
                    }
                }
            }
            gx += __step_3
        }
        gy += __step_2
    }
    var frames: [Any] = __pytra_as_list([])
    let __step_11 = __pytra_int(Int64(1))
    var __loop_10 = __pytra_int(Int64(0))
    while ((__step_11 >= 0 && __loop_10 < __pytra_int(steps)) || (__step_11 < 0 && __loop_10 > __pytra_int(steps))) {
        frames = __pytra_as_list(frames); frames.append(render(grid, w, h, cell))
        grid = __pytra_as_list(next_state(grid, w, h))
        __loop_10 += __step_11
    }
    __pytra_noop(out_path, (__pytra_int(w) * __pytra_int(cell)), (__pytra_int(h) * __pytra_int(cell)), frames, [])
    var elapsed: Double = __pytra_float((__pytra_float(__pytra_perf_counter()) - __pytra_float(start)))
    __pytra_print("output:", out_path)
    __pytra_print("frames:", steps)
    __pytra_print("elapsed_sec:", elapsed)
}

@main
struct Main {
    static func main() {
        run_07_game_of_life_loop()
    }
}
